For Issue No - 10388 Commit #10388 : Factorial2 Issue Sorted

ashutoshsaboo opened the pull request:
Fixes #10388 . Previously, it ran into an infinite loop for fractions and negative integers, Now after this changes, it returns the complex infinity - zoo for integers less than -1, and returns an error message for fractions. So, the problem mentioned in the issue #10388 is basically solved. It can be verified after executing the changed code, with different inputs.

jksuom commented:
I looks like you have based the PR on a fairly old version of SymPy. It may be best to pull an up-to-date version first and then start afresh. As to the values of factorial2 for negative integers, not all of them need be infinite. It is possible to define finite values for negative odd arguments by a recurrence relation or rewrite formula.

ashutoshsaboo committed:
Updated Code with new version of sympy - Fixes Issue No 10388

jksuom commented on sympy/functions/combinatorial/factorials.py:
```diff
@@ -339,13 +339,18 @@ def eval(cls, arg):
+                return arg * (S.NegativeOne) ** ((1 - arg) / 2) / factorial2(-arg)
+            raise ValueError("argument must be nonnegative integer or odd integer")
+        else:
+            raise ValueError("argument must be nonnegative integer or odd integer")
```
This line should not be added. It will be executed, for example, when arg is a symbol. This is wrong because the symbol may be intended to denote an integer. Instead of raising ValueError the method should simply return (with implied None).

ashutoshsaboo committed:
Updated Code for Issue No 10388 - ashutosh.saboo96@gmail.com

ashutoshsaboo commented:
@jksuom Thanks for your comment. I changed the code. Any other problems you see in the code, you can help me out with?

jksuom commented on sympy/functions/combinatorial/factorials.py:
```diff
@@ -339,13 +339,18 @@ def eval(cls, arg):
+                return arg * (S.NegativeOne) ** ((1 - arg) / 2) / factorial2(-arg)
+            raise ValueError("argument must be nonnegative integer or odd integer")
+        else:
+            return None
```
This is not needed (and neither is ``else``). Python will return at the end of a function even if there is no explicit ``return`` statement, and, in such a case, the return value will be ``None``.

jksuom commented on sympy/functions/combinatorial/factorials.py:
```diff
@@ -339,13 +339,18 @@ def eval(cls, arg):
                 raise ValueError("argument must be nonnegative or odd")
 
-            return arg * (S.NegativeOne) ** ((1 - arg) / 2) / factorial2(-arg)
+            if arg.is_integer:
```
I think this test could be earlier (in the form ``arg.is_Integer`` which is more efficient). Probably this test above
```
            if arg.is_infinite:
                return
```
could be replaced with
```
            if not arg.is_Integer:
                raise ValueError("argument must be an integer")
```
Then this ``if arg.is_integer``, and also the ``if arg.is_odd`` bit above could be removed to simplify the code.

ashutoshsaboo replied:
@jksuom Thank you for your help. But the thing is that, We can't replace arg.is_infinite with either of not arg.is_Integer or not arg.is_integer , because , if you initialize a symbol 'x' with no other additional parameters, and no values assigned as well, then always not arg.is_Integer or not arg.is_integer be 'true' and hence it'll enter the first loop, and raise a VaueError instead of returning None. This was the same problem happening in my previous commit, i.e, the commit previous to this one, because I was raising a ValueError in the last 'else' statement, where I am returning None now. That was causing a lot of tests to fail in the Travis CI build. Although yes, as you told the last 'else' statement is not required, I agree to that. Python will itself return none, if nothing is specified. 

So, I feel we can't replace arg.is_infinite with the other two. So, could you help me out as to why the test is failing in the Travis CI Build. There's only one test which is failing. I am not able to figure out, as to why it's failing. :( 
@gxyd @Shekharrajak @asmeurer

ashutoshsaboo committed:
Made code more efficient, Solves Issue no 10388 - ashutosh.saboo96@gmail.com

jksuom commented:
Was there some problem why replacing ``if arg.is_infinite`` with a test of ``arg.is_Integer`` would not work?

ashutoshsaboo commented:
@jksuom I didn't try making the change and committing it. I thought that might be the problem, because in python, I initialized a normal symbol 'x' , without any parameter, and initial value, And I executed "not x.arg.is_integer" and "not x.arg.is_Integer" which both printed true, and hence it would enter the first condition, and raise a ValueError, instead of returning None. Still, I'll try uploading that once, after my current commit's tests get completed. 

Any other problem, you think in the code, that could help me sort out the failing test here - https://travis-ci.org/sympy/sympy/builds/105338820 ? @asmeurer @gxyd @Shekharrajak

ashutoshsaboo commented:
@gxyd So, My all tests have passed, and it solves the issue #10388 ! So, Will you merge this into master? Or do I need to merge it?

jksuom commented:
> And I executed "not x.arg.is_integer" and "not x.arg.is_Integer" which both printed true, and hence it would enter the first condition, and raise a ValueError, instead of returning None.

I think it would enter the first condition (the one corresponding to the current ``if arg.is_infinite``) only if it would pass the test ``if arg.is_Number`` on a previous line. Would your `x` satisfy that?

jksuom commented:
Also, a test of the original issue should be added to sympy/functions/combinatorial/tests/test_comb_factorial.py with a reference to the issue.

ashutoshsaboo commented:
@jksuom Ohh then I totally misinterpreted your point. I was putting the if not arg.is_integer condition in place of the first arg.is_infinite condition, outside the arg.is_Number condition. Oops! Sorry for the inconvenience! 

Could you elaborate as to how do I add that new test? I am new to Open source, hence don't know much about tests. And also, how to merge this commit to the master?

jksuom commented:
An issue is often dedicated a separate test function ``def test_issue_xxxx()`` which would contain the line
```
    raises(ValueError, lambda: factorial2(S(5)/2))
```
(In this case one could also think of adding this line into ``test_factorial2`` (with a reference to the issue) as it really is just a test of the intended behaviour of the function.)

Finally, you don't have to worry about merging to the master. Somebody will take care of that.

ashutoshsaboo committed:
Updated Test for Issue No 10388 | ashutosh.saboo96@gmail.com

ashutoshsaboo committed:
Updated Test File for Issue No 10388 | ashutosh.saboo96@gmail.com

ashutoshsaboo commented:
@jksuom Thanks for your help! I added a comment line in the tests file, writing the issue no which it solves. Also, I added the line `    raises(ValueError, lambda: factorial2(S(5)/2))
` 
Do I need to add anything more? So, Will this get merged now with the master?

jksuom commented:
Currently we get
```
>>> factorial2(oo)
factorial2(oo)
```
This should raise a ValueError because ``oo`` is not an integer. It is caused by the ``if arg.is_infinite`` test. This should also be fixed and a test added.

ashutoshsaboo commented:
@jksuom Thanks for your comment. Since I am new to SymPy, just wanted to know if we represent infinity by 'oo' (i.e, 2 small o's)? Or what does 'oo' represent?

ashutoshsaboo committed:
Updated Code for Issue 10388| raises Error for oo | ashutosh.saboo96@gmail.com

jksuom commented on sympy/functions/combinatorial/factorials.py:
```diff
@@ -331,7 +331,7 @@ def eval(cls, arg):
         if arg.is_Number:
             if arg.is_infinite:
-                return
+                raise ValueError("argument must be nonnegative integer or odd integer")
```
If the line above were replaced by ``if not arg.is_Integer``, then it would catch the infinities and, in addition, arg would be an Integer henceforth.

jksuom commented on sympy/functions/combinatorial/factorials.py:
```diff
@@ -339,13 +339,14 @@ def eval(cls, arg):
                 if arg.is_even:
                     k = arg / 2
                     return 2 ** k * factorial(k)
+                if arg.is_odd:
```
If arg is now an Integer, this test is not necessary...

jksuom commented on sympy/functions/combinatorial/factorials.py:
```diff
@@ -339,13 +339,14 @@ def eval(cls, arg):
                     return 2 ** k * factorial(k)
+                if arg.is_odd:
+                    return factorial(arg) / factorial2(arg - 1)
+                raise ValueError("argument must be nonnegative integer or odd integer")
```
...and this line could be removed.

ashutoshsaboo replied:
@jksuom Thanks for your help! So, I guess the code is more efficient now, than the one was previous. So, Do I need to add anything more now?

ashutoshsaboo committed:
Optimized the code more | Solves Issue No 10388 | ashutosh.saboo96@gmail.com

ashutoshsaboo commented:
@jksuom Thanks for your help! So, I guess the code is more efficient now, than the one was previous. So, Do I need to add anything more now?

jksuom commented:
I think it looks good now, but let's wait for the tests to finish.

ashutoshsaboo commented:
@jksuom Yes sure! That's fine! :+1: :)

ashutoshsaboo commented:
@jksuom As I told, the is_Integer is causing the tests to fail here. I don't know what's the exact problem though, which was also present when I made my 2nd commit. So, I'll change the code back to the previous one, which was passing all the tests. Fine?

And also, why are the test files failing the build tests? We have just added those two lines, which was for this issue, then why is it failing? Could you help on that?

ashutoshsaboo committed:
Changed code, added valueerror for oo | ashutosh.saboo96@gmail.com

jksuom commented:
There seems to be a syntax error in the test file. A missing parentesis after ``factorial2(oo)``.
You should check your modifications locally by ``bin/test factorial`` before pushing them to Github.

ashutoshsaboo commented:
Thanks @jksuom ! 

These are the 2 lines that I added in the test file-:

```
    raises (ValueError, lambda: factorial2(oo))
    raises (ValueError, lambda: factorial2(S(5)/2))

```
So, I put the missing parenthesis, and then ran from my system `bin/test factorial` . Sadly it gave me this output-:

```
sympy/functions/combinatorial/tests/test_comb_factorials.py[13] ...f...E..... 
                                                                          [FAIL]

________________________________________________________________________________
_ sympy/functions/combinatorial/tests/test_comb_factorials.py:test_factorial2 __
  File "/home/ashutoshsaboo/SymPyMain/sympy/sympy/functions/combinatorial/tests/test_comb_factorials.py", line 199, in test_factorial2
    raises (ValueError, lambda: factorial2(oo))
NameError: global name 'raises' is not defined

= tests finished: 11 passed, 1 expected to fail, 1 exceptions, in 1.63 seconds =
DO *NOT* COMMIT!

```

So, How do I fix this? Could you help me out with this?

leosartaj commented:
@ashutoshsaboo You need to import ``raises``. Probably on line-6.

ashutoshsaboo committed:
Updated Test for Issue No 10388 | ashutosh.saboo96@gmail.com

jksuom commented:
Did you forget ``is_Integer``? That was probably not the reason for the failure.

ashutoshsaboo commented:
@leosartaj Thanks ! :)

@jksuom I'll wait for this new commit's build tests to finish. I'll check that once again and try pushing the new commit again.

ashutoshsaboo committed:
Efficiency of Code increased | Issue No 10388 | ashutosh.saboo96@gmail.com

ashutoshsaboo commented:
@jksuom So, now it has passed all the tests. Can this be merged now into the master?

jksuom commented:
I think it can be merged now. Thanks!
